research/fed-sm/job_configs/fedsm_prostate/app/custom/workflows/scatter_and_gather_fedsm.py

```diff
@@ -0,0 +1,255 @@
+        - persistor_fedsm is customized for persisting FedSM model set (a dict of models) for all clients
+        Each client sends it's updated three weights after local training, to be aggregated accordingly by aggregator,
+        we use one customized aggregator to handle all three models,
+        global and selector models following standard weighted average, while personalized models following SoftPull
+        The shareable generator is used to convert the aggregated weights to shareable, and shareable back to weights.
+
+        Args:
+            client_id_label_mapping: needed for training selector model, no Default.
+            min_clients (int, optional): Min number of clients in training. Defaults to 1.
+            num_rounds (int, optional): The total number of training rounds. Defaults to 5.
+            start_round (int, optional): Start round for training. Defaults to 0.
+            wait_time_after_min_received (int, optional): Time to wait before beginning aggregation after contributions received. Defaults to 10.
+            train_timeout (int, optional): Time to wait for clients to do local training.
+            aggregator_id (str, optional): ID of the aggregator component for FedSM models. Defaults to "aggregator".
+            persistor_id_fedsm (str, optional): ID of the persistor component for FedSM models. Defaults to "persistor_fedsm".
+            shareable_generator_id (str, optional): ID of the shareable generator. Defaults to "shareable_generator".
+            train_task_name (str, optional): Name of the train task. Defaults to "train".
+        """
+
+        super().__init__(
+            min_clients=min_clients,
+            num_rounds=num_rounds,
+            start_round=start_round,
+            wait_time_after_min_received=wait_time_after_min_received,
+            aggregator_id=aggregator_id,
+            persistor_id=persistor_id,
+            shareable_generator_id=shareable_generator_id,
+            train_task_name=train_task_name,
+            train_timeout=train_timeout,
+            ignore_result_error=ignore_result_error,
+        )
+        # extras for FedSM
+        # client_id to label mapping for selector
+        self.client_id_label_mapping = client_id_label_mapping
+
+    def start_controller(self, fl_ctx: FLContext) -> None:
+        super().start_controller(fl_ctx=fl_ctx)
+        self.log_info(fl_ctx, "Initializing FedSM-specific workflow components.")
+        self.log_info(
+            fl_ctx,
+            "Client_ID to selector label mapping: {}".format(self.client_id_label_mapping),
+        )
+        # get engine
+        engine = fl_ctx.get_engine()
+        if not engine:
+            self.system_panic("Engine not found. ScatterAndGather exiting.", fl_ctx)
+            return
+
+        # Get all clients
+        clients = engine.get_clients()
+        self.participating_clients = [c.name for c in clients]
```

Comment:
should be initialized with None in __init__()

Reply:
added